src/main.rs

```diff
@@ -2,34 +2,60 @@ use rand_core::{CryptoRng, OsRng, RngCore};
 use std::{env, time};
 
 use frost::frost::{Party, PolyCommitment, PublicNonce, SignatureAggregator, SignatureShare};
+use secp256k1_math::{point::Point, scalar::Scalar};
 
-use hashbrown::HashMap;
+use hashbrown::{HashMap, HashSet};
+
+// pulls out on the shares relevant to party_id
+// from all the shares that were broadcast
+fn filter_party_shares(
+    party_id: usize,
+    key_owners: &Vec<usize>,
+    broadcast_shares: &Vec<Vec<(usize, Scalar)>>,
+) -> HashMap<usize, Vec<(usize, Scalar)>> {
+    let mut shares: HashMap<usize, Vec<(usize, Scalar)>> = HashMap::new();
+
+    for sender in 0..broadcast_shares.len() {
+        for (key_id, s) in &broadcast_shares[sender] {
+            if party_id == key_owners[*key_id] {
+                // if shares[key_id] doesn't exist create a new vector
+                if shares.contains_key(key_id) == false {
+                    shares.insert(key_id.clone(), Vec::new());
+                }
+
+                let v = shares.entry(*key_id).or_insert(Vec::new());
+                v.push((sender, s.clone()));
+            }
+        }
+    }
+    shares
+}
 
 // This will eventually need to be replaced by rpcs
 #[allow(non_snake_case)]
 fn distribute(
     parties: &mut Vec<Party>,
+    key_owners: &Vec<usize>, // N-long vector with indices = key_id and values = party_id
     A: &Vec<PolyCommitment>,
     B: &Vec<Vec<PublicNonce>>,
-) -> u128 {
+) -> (u128, HashMap<usize, Point>) {
     // each party broadcasts their commitments
-    // these hashmaps will need to be serialized in tuples w/ the value encrypted
+    // these will need to be serialized in tuples w/ the value encrypted
     let mut broadcast_shares = Vec::new();
-    for i in 0..parties.len() {
-        broadcast_shares.push(parties[i].get_shares());
+    for party_id in 0..parties.len() {
+        broadcast_shares.push(parties[party_id].get_shares());
     }
 
+    let mut public_keys = HashMap::new();
     let mut total_compute_secret_time = 0;
-
-    // each party collects its shares from the broadcasts
-    // maybe this should collect into a hashmap first?
-    for i in 0..parties.len() {
-        let mut h = HashMap::new();
-        for j in 0..parties.len() {
-            h.insert(j, broadcast_shares[j][&i]);
-        }
+    for party_id in 0..parties.len() {
+        let party_shares = filter_party_shares(party_id, &key_owners, &broadcast_shares);
         let compute_secret_start = time::Instant::now();
-        parties[i].compute_secret(h, &A);
+        let mut pks = parties[party_id].compute_secret(party_shares, &A);
+        for (k, v) in pks.drain() {
+            // TODO: Is there a better way to add hashmaps?
+            public_keys.insert(k, v);
+        }
```

Comment:
Yes, in this case I'd recommend using the `extend` method. We also don't need to call `.drain()` since we want to consume `pks` and HashMap implements `IntoIterator`. (Assuming hashbrowns HashMap works the same way as the one in std).
```suggestion
        public_keys.extend(pks);
```